gdal/frmts/grib/gribcreatecopy.cpp

```diff
@@ -749,17 +791,35 @@ float* GRIB2Section567Writer::GetFloatData()
     }
     CPLErr eErr = m_poSrcDS->GetRasterBand(m_nBand)->RasterIO(
         GF_Read,
-        0, 0,
-        m_nXSize, m_nYSize,
+        m_nSplitAndSwap, 0,
+        m_nXSize - m_nSplitAndSwap, m_nYSize,
         pafData + (m_adfGeoTransform[5] < 0 ? (m_nYSize - 1) * m_nXSize : 0),
-        m_nXSize, m_nYSize,
+        m_nXSize - m_nSplitAndSwap, m_nYSize,
         GDT_Float32,
         sizeof(float),
         m_adfGeoTransform[5] < 0 ?
             -static_cast<GSpacing>(m_nXSize * sizeof(float)):
             static_cast<GSpacing>(m_nXSize * sizeof(float)),
         nullptr);
-    if( eErr != CE_None )
+    if (eErr != CE_None)
+    {
+        VSIFree(pafData);
+        return nullptr;
+    }
+    eErr = m_poSrcDS->GetRasterBand(m_nBand)->RasterIO(
```

Comment:
this could be conditionalized to m_nSplitAndSwap > 0 (otherwise RasterIO() will emit a debug message about an odd window being passed)